presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java

```diff
@@ -93,7 +96,38 @@ public PlanNode extract()
 
     private PlanNode extract(PlanNode node)
     {
-        return resolveGroupReferences(node, Lookup.from(planNode -> ImmutableList.of(this.resolve(planNode))));
+        return resolveGroupReferences(node, getLookup());
+    }
+
+    public Lookup getLookup()
+    {
+        return new Lookup()
+        {
+            @Override
+            public List<PlanNode> resolveGroup(PlanNode node)
+            {
+                checkArgument(node instanceof GroupReference, "node is not a group reference");
+                return ImmutableList.of(membership.get(((GroupReference) node).getGroupId()));
+            }
+
+            @Override
+            public TraitSet resolveTraitSet(PlanNode node)
+            {
+                if (node instanceof GroupReference) {
+                    GroupReference groupReference = (GroupReference) node;
+                    return groupTraits.getOrDefault(groupReference.getGroupId(), TraitSet.empty());
+                }
+                else {
+                    // TODO: REMOVE THIS LOOP BEFORE PULL REQUEST MERGE
+                    for (Map.Entry<Integer, PlanNode> entry : membership.entrySet()) {
```

Comment:
I think you need to maintain reverse mapping from `planNode -> groupId`. You can use http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/BiMap.html to maintain consistency and avoid bugs (`membership` field).

Reply:
Yeah... I know could do such things. But I used to have a map where `PlanNode` was a key and you did not like it.

Reply:
`BiMap` preserves the mapping from `planNode` to `group`. You can still have `BiMap<Integer, PlanNode> membership`.